### PR TITLE
[MOB-1757] Integrate HTTP server on IOS driver works locally

### PR DESCRIPTION
**Pull request by amanjeetsingh150**
## Proposed Changes
1. Integrate HTTP server requests in `IOSDriver`
2. Refactor `Maestro.ios` to take `IOSDriver` as input
3. Ensure xcuitest artifacts install while setting up IOSDriver
4. Adds logging so that we can root cause issues
5. Search elements from node children depth first 

## Testing

https://user-images.githubusercontent.com/12881364/206658944-f0cc5c51-d626-40b5-a3b9-3fe209600ca9.mp4

### CODE REVIEW

File: maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt [hunk 1]
```diff
@@ -122,19 +115,7 @@ object DeviceService {
                 process
                     .waitFor(1000, TimeUnit.MILLISECONDS)
                 process.exitValue() == 0
-            } || error("Simulator failed to boot")
```
Comment (dmitry-zaitsev) [inline maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt:125]: This should stay, I believe

File: maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt [hunk 2]
```diff
@@ -122,19 +115,7 @@ object DeviceService {
...
-
-            // Test if idb can get accessibility info elements with non-zero frame with
-            logger.warning("Waiting for Accessibility info to become available..")
-            MaestroTimer.retryUntilTrue(timeoutMs = 20000, delayMs = 100) {
```
Comment (dmitry-zaitsev) [inline maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt:129]: We still need to wait for IDB to boot, otherwise there is a risk that `tap` command will fail because IDB is not ready. We can't use `contentDescriptor` for that anymore, so we need some other way to deduce that

File: maestro-client/build.gradle [hunk 3]
```diff
@@ -65,6 +65,13 @@ dependencies {
     api 'org.mozilla:rhino:1.7.14'
     api 'org.jsoup:jsoup:1.15.3'
     api 'com.squareup.okhttp3:okhttp:4.10.0'
+    api 'org.rauschig:jarchivelib:1.2.0'
+    api 'com.fasterxml.jackson.core:jackson-databind:2.13.4'
+    api 'com.fasterxml.jackson.module:jackson-module-kotlin:2.13.4'
+    api 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.13.4'
+    api 'com.fasterxml.jackson.dataformat:jackson-dataformat-xml:2.13.4'
+    api 'commons-io:commons-io:2.11.0'
```
Comment (dmitry-zaitsev) [inline maestro-client/build.gradle:73]: What do we need commons-io for? Kotlin typically has many of the common IO utils, so just double checking whether we really need it

File: maestro-client/src/main/java/maestro/Maestro.kt [hunk 4]
```diff
@@ -437,16 +437,7 @@ class Maestro(private val driver: Driver) : AutoCloseable {
...
-            val channel = ManagedChannelBuilder.forAddress(host, port)
-                .usePlaintext()
-                .build()
-
-            return ios(channel)
-        }
-
-        fun ios(channel: ManagedChannel): Maestro {
-            val driver = IOSDriver(IdbIOSDevice(channel))
+        fun ios(driver: Driver): Maestro {
```
Comment (dmitry-zaitsev) [inline maestro-client/src/main/java/maestro/Maestro.kt:440]: Maestro already have a constructor that takes `Driver` as a parameter. This is not much different except that it calls `driver.open()`

Why not have `ios()` method actually create a driver internally?

File: maestro-client/src/main/java/maestro/MaestroTimer.kt [hunk 5]
```diff
@@ -1,5 +1,7 @@
 package maestro
 
+import java.util.concurrent.TimeoutException
```
Comment (dmitry-zaitsev) [inline maestro-client/src/main/java/maestro/MaestroTimer.kt:3]: Does not seem to be used?

File: maestro-client/src/main/java/maestro/drivers/IOSDriver.kt [hunk 6]
```diff
@@ -45,12 +51,39 @@ class IOSDriver(
...
+
+        logger.info("[Start] Ensuring ui test runner app is launched on ${iosDevice.deviceId}")
+        IOSUiTestRunner.ensureOpen()
+        logger.info("[Done] Ensuring ui test runner app is launched on ${iosDevice.deviceId}")
+
+        logger.info("[Start] Trying to view hierarchy for ${iosDevice.deviceId}")
+        val nodes = iosDevice
+            .contentDescriptor(appId = IOSUiTestRunner.UI_TEST_RUNNER_APP_BUNDLE_ID)
+            .get()
+        MaestroTimer.retryUntilTrue(5000) {
```
Comment (dmitry-zaitsev) [inline maestro-client/src/main/java/maestro/drivers/IOSDriver.kt:80]: You are retrying a static piece of code here, should you call `contentDescriptor` inside of the `retry` block?

File: maestro-client/src/main/java/maestro/drivers/IOSDriver.kt [hunk 7]
```diff
@@ -59,9 +92,12 @@ class IOSDriver(
 
     override fun close() {
         iosDevice.close()
+        iosDevice.uninstall(IOSUiTestRunner.UI_TEST_RUNNER_APP_BUNDLE_ID)
```
Comment (dmitry-zaitsev) [inline maestro-client/src/main/java/maestro/drivers/IOSDriver.kt:95]: I think we need to uninstall it before installing it again. There is no guarantee that `close()` will always be called

File: maestro-client/src/main/java/maestro/ios/IOSUiTestRunner.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        val processOutput = ProcessBuilder(
+            "bash",
+            "-c",
+            "xcrun simctl spawn booted launchctl print system | grep $UI_TEST_RUNNER_APP_BUNDLE_ID | awk '/$UI_TEST_RUNNER_APP_BUNDLE_ID/ {print \$3}'"
+        ).start().inputStream.source().buffer().readUtf8().trim()
+
+        if (!processOutput.contains(UI_TEST_RUNNER_APP_BUNDLE_ID)) {
+            logger.info("Not able to find ui test runner app, going to install now")
+            val tempDir = System.getenv("TMPDIR")
+            val xctestRunFile = File("$tempDir/maestro-driver-ios-config.xctestrun")
```
Comment (dmitry-zaitsev) [inline maestro-client/src/main/java/maestro/ios/IOSUiTestRunner.kt:35]: nit: you can do `File(tempDir, "maestro-driver-ios-config.xctestrun")` instead

File: maestro-ios/src/main/java/ios/hierarchy/Frame.kt [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package ios.hierarchy
+
+import com.google.gson.annotations.SerializedName
+
+data class Frame(
+    @SerializedName("Width") val Width: Float,
```
Comment (dmitry-zaitsev) [inline maestro-ios/src/main/java/ios/hierarchy/Frame.kt:6]: Since you are already using `SerializedName`, let's keep actual field names canonical (camelCase)

File: maestro-test/src/test/resources/038_partial_id.yaml [hunk 10]
```diff
@@ -4,5 +4,5 @@ appId: com.example.app
     id: "keyboard_area"
     retryTapIfNoChange: false
 - tapOn:
-    id: ".*keyboard_area"
+    id: ".*another_keyboard_area"
```
Comment (dmitry-zaitsev) [inline maestro-test/src/test/resources/038_partial_id.yaml:7]: Why was this changed?

File: maestro-client/src/main/resources/maestro-driver-ios-config.xctestrun [hunk 11]
```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (artem888) [inline maestro-client/src/main/resources/maestro-driver-ios-config.xctestrun:1]: long-term should those files be under the source control? for me they can be stored elsewhere, i.e. gcloud

File: maestro-driver-ios/build-maestro-runner.sh [hunk 12]
```diff
@@ -0,0 +1,20 @@
+## Build the UI test
```
Comment (artem888) [inline maestro-driver-ios/build-maestro-runner.sh:1]: whey there are two files:
- maestro-driver-ios/build-maestro-runner.sh
- maestro-driver-ios/build-maestro-ios-runner.sh

File: maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt (hunk 2 above)
Comment (amanjeetsingh150) [reply maestro-cli/src/main/java/maestro/cli/device/DeviceService.kt:129]: True that I have used tap result to check if idb is ready. Dispatching tap event on 0,0 coordinates and see if it's Ok.

File: maestro-client/build.gradle (hunk 3 above)
Comment (amanjeetsingh150) [reply maestro-client/build.gradle]: Removed this and used File.deleteRecursively

File: maestro-client/src/main/java/maestro/MaestroTimer.kt (hunk 5 above)
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/MaestroTimer.kt]: Removed ✅

File: maestro-client/src/main/java/maestro/drivers/IOSDriver.kt (hunk 6 above)
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/drivers/IOSDriver.kt:77]: Good catch have replaced with call to contentdescriptor

File: maestro-ios/src/main/java/ios/hierarchy/Frame.kt (hunk 9 above)
Comment (amanjeetsingh150) [reply maestro-ios/src/main/java/ios/hierarchy/Frame.kt]: Done, changed!

File: maestro-client/src/main/java/maestro/ios/Simctl.kt [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import maestro.MaestroTimer
+import ios.commands.CommandLineUtils
+import okio.buffer
+import okio.source
+import java.util.concurrent.TimeUnit
+
+object Simctl {
```
[the remark this answers is not in the document]
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/ios/Simctl.kt:11]: Yup have changed path of the same file to be inside maestro-client

File: maestro-test/src/test/resources/038_partial_id.yaml (hunk 10 above)
Comment (amanjeetsingh150) [reply maestro-test/src/test/resources/038_partial_id.yaml:7]: This was changed because we now match from depth first due to this change: https://github.com/mobile-dev-inc/maestro/pull/477/commits/bdad3e52090e95820adcfb4737331eb8b9552e87

File: maestro-driver-ios/build-maestro-runner.sh (hunk 12 above)
Comment (amanjeetsingh150) [reply maestro-driver-ios/build-maestro-runner.sh]: Removed one

File: maestro-client/src/main/resources/maestro-driver-ios-config.xctestrun (hunk 11 above)
Comment (amanjeetsingh150) [reply maestro-client/src/main/resources/maestro-driver-ios-config.xctestrun:1]: I think its fine to keep it here since these are intentionally packaged in jar.

File: maestro-client/src/main/java/maestro/drivers/IOSDriver.kt (hunk 7 above)
Comment (amanjeetsingh150) [reply maestro-client/src/main/java/maestro/drivers/IOSDriver.kt]: Done

File: maestro-client/src/main/java/maestro/Maestro.kt (hunk 4 above)
Comment (dmitry-zaitsev) [reply maestro-client/src/main/java/maestro/Maestro.kt:440]: `Maestro` already has a constructor that takes `Driver` as a parameter, we can just use that constructor for Cloud

File: maestro-test/src/test/resources/038_partial_id.yaml (hunk 10 above)
Comment (dmitry-zaitsev) [reply maestro-test/src/test/resources/038_partial_id.yaml:7]: This doesn't seem right in this test case:

```
val driver = driver {
            element {
                id = "com.google.android.inputmethod.latin:id/another_keyboard_area"
                bounds = Bounds(0, 0, 100, 100)
            }

            element {
                id = "com.google.android.inputmethod.latin:id/keyboard_area"
                bounds = Bounds(0, 100, 100, 200)
            }
}
```

Those views are siblings, so I would've expected `id: .*keyboard_area` to match the first view

Comment (amanjeetsingh150) [reply maestro-test/src/test/resources/038_partial_id.yaml]: FIxed ✅
